"Look Up" and "Search Web" should appear in the edit menu for selected text

whsieh opened the pull request:
#### f778680e9ba45d5bbcd17b16dd4e32886370e73e
<pre>
&quot;Look Up&quot; and &quot;Search Web&quot; should appear in the edit menu for selected text
<a href="https://bugs.webkit.org/show_bug.cgi?id=285086">https://bugs.webkit.org/show_bug.cgi?id=285086</a>
<a href="https://rdar.apple.com/141917634">rdar://141917634</a>

Reviewed by Abrar Rahman Protyasha.

Add support for a few more text services: &quot;Look Up&quot; and &quot;Search Web&quot;, through the edit menu for
selected text in PDFs when Unified PDF is enabled. See below for more details.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::stringsBeforeAndAfterSelection const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::documentEditingContext const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::stringsBeforeAndAfterSelection const):

Add a new helper method to surface text (i.e. a given number of characters) around the current
selection, as a `std::pair` representing text before and after (respectively).

(WebKit::UnifiedPDFPlugin::documentEditingContext const):

Add a helper method to return a document editing context for the given request. This only surfaces
selected text at the moment, but can be fleshed out in the future if needed.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::stringsBeforeAndAfterSelection const):
(WebKit::PluginView::documentEditingContext const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getSelectionContext):

Implement this for the unified PDF plugin in terms of the new `stringsBeforeAndAfterSelection`
helper method. Look Up uses this to obtain text context around the selection string.

(WebKit::WebPage::requestDocumentEditingContext):

Add very basic support for document editing contexts in PDFs — DataDetectorsUI uses a Reveal item
derived from the document editing context to determine whether or not they should populate the edit
menu with &quot;Look Up&quot; and &quot;Search Web&quot;. We currently return an empty document editing context, which
causes this to fail.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
(-[TestWKWebView synchronouslyRequestDocumentContext:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Add a new API test to exercise this change by simulating the &quot;Look Up&quot; action from API.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView synchronouslyRequestDocumentContext:]):

Move this testing helper out of the test file and into `TestWKWebView`, for use in Unified PDF
tests as well.

Canonical link: <a href="https://commits.webkit.org/288247@main">https://commits.webkit.org/288247@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9a6c7b1035dc05941de08339605d5317c213b098

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82228 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1892 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36289 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87360 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33289 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84333 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/1963 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9772 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64083 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21833 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85297 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1356 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74816 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44361 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1256 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/28998 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32330 "Built successfully") | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29619 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88716 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9534 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6792 "Found 1 new test failure: http/wpt/mediarecorder/record-96KHz-sources.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72474 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9760 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70631 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71692 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17880 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15826 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/14849 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/886 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9487 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/14977 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9361 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12827 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11130 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
```diff
@@ -28,6 +28,7 @@
 #if ENABLE(PDF_PLUGIN)
 
 #include "CursorContext.h"
+#include "DocumentEditingContext.h"
```
Nit: I think you can forward decl `DocumentEditingContext[Request]` here

aprotyas replied:
Oh, I see, can't happen because of the default ctor invoked for `DocumentEditingContext`.

whsieh replied:
Oh, I can move the method stub out into the implementation, and then just use forward declarations in the headers.

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -4203,6 +4245,43 @@ static bool isEmpty(PDFSelection *selection)
 
+DocumentEditingContext UnifiedPDFPlugin::documentEditingContext(DocumentEditingContextRequest&& request) const
+{
+    static constexpr auto unsupportedOptions = OptionSet {
```
Nit: maybe fewer characters typed this way?

```suggestion
    static constexpr OptionSet unsupportedOptions {
```

aprotyas replied:
Also, another nit, if you want the enum class benefits but don't want to type `DocumentEditingContextRequest::Options::` over and over in this method...

```
using enum DocumentEditingContextRequest::Options;

// request.options.contains(AttributedText);
```

whsieh replied:
Will do!

whsieh committed:
"Look Up" and "Search Web" should appear in the edit menu for selected text
https://bugs.webkit.org/show_bug.cgi?id=285086
rdar://141917634

Reviewed by Abrar Rahman Protyasha.

Add support for a few more text services: "Look Up" and "Search Web", through the edit menu for
selected text in PDFs when Unified PDF is enabled. See below for more details.

* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
(WebKit::PDFPluginBase::stringsBeforeAndAfterSelection const):
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::documentEditingContext const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::stringsBeforeAndAfterSelection const):

Add a new helper method to surface text (i.e. a given number of characters) around the current
selection, as a `std::pair` representing text before and after (respectively).

(WebKit::UnifiedPDFPlugin::documentEditingContext const):

Add a helper method to return a document editing context for the given request. This only surfaces
selected text at the moment, but can be fleshed out in the future if needed.

* Source/WebKit/WebProcess/Plugins/PluginView.cpp:
(WebKit::PluginView::stringsBeforeAndAfterSelection const):
(WebKit::PluginView::documentEditingContext const):
* Source/WebKit/WebProcess/Plugins/PluginView.h:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::getSelectionContext):

Implement this for the unified PDF plugin in terms of the new `stringsBeforeAndAfterSelection`
helper method. Look Up uses this to obtain text context around the selection string.

(WebKit::WebPage::requestDocumentEditingContext):

Add very basic support for document editing contexts in PDFs — DataDetectorsUI uses a Reveal item
derived from the document editing context to determine whether or not they should populate the edit
menu with "Look Up" and "Search Web". We currently return an empty document editing context, which
causes this to fail.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/DocumentEditingContext.mm:
(-[TestWKWebView synchronouslyRequestDocumentContext:]): Deleted.
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UnifiedPDFTests.mm:
(TestWebKitAPI::UNIFIED_PDF_TEST):

Add a new API test to exercise this change by simulating the "Look Up" action from API.

* Tools/TestWebKitAPI/cocoa/TestWKWebView.h:
* Tools/TestWebKitAPI/cocoa/TestWKWebView.mm:
(-[WKWebView synchronouslyRequestDocumentContext:]):

Move this testing helper out of the test file and into `TestWKWebView`, for use in Unified PDF
tests as well.

Canonical link: https://commits.webkit.org/288247@main
